aten/src/TH/THTensor.hpp

```diff
@@ -97,32 +101,83 @@ struct THTensor
         delete this;
       }
     }
+
+    friend const int64_t* THTensor_getSizePtr(THTensor* tensor);
+    friend const int64_t* THTensor_getStridePtr(THTensor* tensor);
+    friend void THTensor_resizeDim(THTensor* tensor, int64_t ndim);
+    friend void THTensor_setSizesAndStrides(THTensor* tensor, std::vector<int64_t>&& new_size, std::vector<int64_t>&& new_stride);
+    friend void THTensor_setSizeAtDim(THTensor* tensor, int dim, int64_t new_size);
+    friend void THTensor_setStrideAtDim(THTensor* tensor, int dim, int64_t new_stride);
+    friend void THTensor_setStorageOffset(THTensor* tensor, ptrdiff_t storage_offset);
```

Comment:
I don't really get this friend strategy.  If it's a friend, it's part of the API, why not just make it a method?

Reply:
Also can we just get rid of some of these?  In particular it's really strange to need to call THTensor_getStoragePtr -- there's no Storage method on THTensor?  (I know that's probably because we are moving storages around, but there's almost certainly an equivalent in C10, right?)

Reply:
It would be nice to have a little audit of where these are actually called.

Reply:
I didn't want to get rid of `THTensor_getStoragePtr` because I don't want to return raw storage pointer in the "official" API, because after @cpuhrsch's patch merges I can return an `at::Storage` instead, and that's a lot safer.

Reply:
Mini-audit (I did TH only):

THTensor_resizeDim
* Why isn't this in the official API? It's semantics are a bit wonky: if you reduce the dimension, we truncate trailing dimensions (in general, this will result in a non-contiguous tensor before you reset the strides); if you increase the dimension, the result used to be undefined (now, each added dimension has size and stride 1).
* Call-sites:
  * Narrow: the function sets sizes and strides, and reduces dim by one. We could probably replace this with `THTensor_setSizesAndStrides`
  * Squeeze: the function sets sizes and strides, then reduces the dim. Replace with `THTensor_setSizesAndStrides`
  * Select: the function sets sizes and strides, then reduces dim. Replace with `THTensor_setSizesAndStrides`.
  * Unsqueeze: the function increases the number of dimensions by one, and then writes in the correct sizes/strides afterwards. Replace with `THTensor_setSizesAndStrides`
  * Resize: the function changes the number of dimensions to the requested amount, then fills in correct sizes/strides. Replace with `THTensor_setSizesAndStrides`
* Conclusion: Replace this with `THTensor_setSizesAndStrides`

THTensor_setSizesAndStrides
* Why isn't this in the official API? It is a function that allows invariant breaking: you can easily set the sizes and strides so that they go out of bounds of the allocated storage, leading to a subsequent segfault. Maybe this is not a good reason and we should make it a method.
* Call-sites:
  * Unfold: Used to restride the tensor
  * (see above for functions we probably will replace with this)

THTensor_setSizeAtDim/THTensor_setStrideAtDim
* Why isn't this in the official API? Like setSizesAndStrides, it can be used to break invariants.
* Call-sites:
  * These two functions seem to always be used together
  * It's generally used anywhere resizeDim was called: narrow, select, squeeze, unsqueeze, resize
  * Transpose: Used to swap the sizes and strides at the transposed dimension
  * Max/min/sum/prod/logicalAnd/logicalAny: Used as a poor mans "expand_as" (when ported to ATen, probably just use `expand_as` directly)

THTensor_setStorageOffset
* Why isn't this in the official API? Invariant breaking. Also, it's unclear if the offset is bytes or numels (it's numels).
* Call-sites:
  * Narrow: for obvious reasons, though arguably set should also handle setting the storage offset as well
  * SetStorageNd: Allows setting the storage offset

THTensor_stealAndSetStoragePtr
* Why not? It takes a raw pointer; I want it to take a Storage RAII smart pointer so that it can't be misused.
* Call-sites:
  * SetStorageNd: Not actually stealing: we retain the pointer before we take it.
  * ResizeNd: Used to set a fresh storage (so yes, stealing)

Reply:
re:
>> I didn't want to get rid of THTensor_getStoragePtr because I don't want to return raw storage pointer in the "official" API, because after @cpuhrsch's patch merges I can return an at::Storage instead, and that's a lot safer.

that was what I was trying to get across -- just do whatever you need to do and we'll resolve it when we have one Storage/StorageImpl.

Reply:
re: THTensor_setSizesAndStrides

This is just `as_strided`, right?  Which we have in our API.  I'm okay if we have safe/unsafe version or whatever.